### ISSUE TITLE
Nodes failing with OutOfMemoryError after about a week of uptime

### ISSUE THREAD

**Issue by bra-fsn** [elastic/elasticsearch#41337]
**Elasticsearch version** (`bin/elasticsearch --version`): 6.6.2-oss

**Plugins installed**: analysis-icu

**JVM version** (`java -version`): 1.8.0_202

**OS version** (`uname -a` if on a Unix-like system): FreeBSD 11.2-STABLE

**Description of the problem including expected versus actual behavior**: I have a relatively stable cluster of 56 (40 data) nodes. My problem is that after some days/weeks of uptime some nodes start to fail with OOM. These tend to happen during the period, when some scrolls and more updates/deletes are done, but it's not rare outside that window as well.
I've already tried to raise the heap from 8G to 12G, without success.
When I restart the cluster, the nodes work fine for another week/days, so I guess if the heap size would be the sole problem, it should cause instabilities sooner.

**Steps to reproduce**: I don't have a way to reproduce it programmatically. It seems to happen with elapsed time. Mostly, the cluster members are (re)started together and when one node falls out with an OOM, there are cases when several nodes follow it. I could observe these OOMs only after a given uptime (mostly weeks, but one week (10-12 days) without any problems is quite usual).
The cluster usage is quite consistent over time, repeating the same patterns, I can't see any spikes in any monitored metrics before the OOM.
Last month, one node had OOM at 2019-03-19T06:14:15 then another at 2019-03-19T08:30:47. All of them had an uptime of 10.4 days.
I have heap dumps from a number of these occasions (currently 19, some with bigger, some with smaller heaps).
I can provide any of these privately if requested.
Opening the last one in Eclipse MAT and running memory leak suspects yields the following:

> Problem Suspect 1
> 
> 27 instances of "org.apache.lucene.index.IndexWriter", loaded by "sun.misc.Launcher$AppClassLoader @ 0x87c901df8" occupy 1,444,023,192 (17.99%) bytes.
> Biggest instances:
> 
> org.apache.lucene.index.IndexWriter @ 0x8e322cd30 - 354,902,448 (4.42%) bytes.
> org.apache.lucene.index.IndexWriter @ 0x896cc48a8 - 194,821,848 (2.43%) bytes.
> org.apache.lucene.index.IndexWriter @ 0x8dfc52008 - 168,491,352 (2.10%) bytes.
> org.apache.lucene.index.IndexWriter @ 0x8f59fdef8 - 149,564,464 (1.86%) bytes.
> org.apache.lucene.index.IndexWriter @ 0x9157db0b0 - 138,265,992 (1.72%) bytes.
> org.apache.lucene.index.IndexWriter @ 0x89e62dec0 - 137,079,136 (1.71%) bytes.
> org.apache.lucene.index.IndexWriter @ 0x8f424cdc0 - 120,118,608 (1.50%) bytes.
> 

> Problem Suspect 2
> 
> 75,375 instances of "org.apache.lucene.codecs.compressing.CompressingStoredFieldsReader", loaded by "sun.misc.Launcher$AppClassLoader @ 0x87c901df8" occupy 1,335,024,544 (16.63%) bytes. 

> Problem Suspect 3
> 
> 5,010 instances of "byte[]", loaded by "<system class loader>" occupy 992,431,728 (12.36%) bytes. 

Outside of these, 4 GiB is reported in other stuff.

Used heap dump | 7.5 GB
-- | --
Number of objects | 21,550,603
Number of classes | 14,541
Number of class loaders | 75
Number of GC roots | 3,418

Biggest objects:

Class Name | Shallow Heap | Retained Heap
-- | -- | --
org.elasticsearch.indices.IndicesQueryCache$ElasticsearchLRUQueryCache @ 0x8879c3208 » | 88 | 754,786,360
org.apache.lucene.index.IndexWriter @ 0x8e322cd30 » | 208 | 354,902,448
org.apache.lucene.index.IndexWriter @ 0x896cc48a8 » | 208 | 194,821,848
org.elasticsearch.search.SearchService @ 0x8b5c48798 » | 104 | 170,363,752
org.apache.lucene.index.IndexWriter @ 0x8dfc52008 » | 208 | 168,491,352
org.elasticsearch.common.util.PageCacheRecycler @ 0x887b88ac8 » | 32 | 161,909,848
org.apache.lucene.index.IndexWriter @ 0x8f59fdef8 » | 208 | 149,564,464
org.apache.lucene.index.IndexWriter @ 0x9157db0b0 » | 208 | 138,265,992
org.apache.lucene.index.IndexWriter @ 0x89e62dec0 » | 208 | 137,079,136
org.apache.lucene.index.IndexWriter @ 0x8f424cdc0 » | 208 | 120,118,608
org.elasticsearch.index.IndexService @ 0x8b0ed27e8 » | 152 | 91,669,488
Total: 11 entries

Packages:

Package | Retained Heap | Retained Heap, % | # Top Dominators
-- | -- | -- | --
<all>First 10 of 635,805 objects | 8,027,986,976 | 100.00% | 635,805
orgFirst 10 of 472,651 objects | 5,969,889,520 | 74.36% | 472,651
apacheFirst 10 of 130,992 objects | 4,066,876,496 | 50.66% | 130,992
luceneFirst 10 of 129,804 objects | 4,066,572,360 | 50.65% | 129,804
indexFirst 10 of 31,273 objects | 1,875,677,160 | 23.36% | 31,273
IndexWriterFirst 10 of 28 objects | 1,444,048,200 | 17.99% | 28
SegmentReaderFirst 10 of 4,445 objects | 391,440,960 | 4.88% | 4,445
Total: 2 entries | 1,835,489,160 |   | 4,473
codecsFirst 10 of 86,643 objects | 1,845,488,368 | 22.99% | 86,643
compressingFirst 10 of 79,011 objects | 1,424,856,096 | 17.75% | 79,011
CompressingStoredFieldsReaderFirst 10 of 75,376 objects | 1,335,025,368 | 16.63% | 75,376
CompressingStoredFieldsIndexReaderFirst 10 of 3,610 objects | 89,664,576 | 1.12% | 3,610
Total: 2 entries | 1,424,689,944 |   | 78,986
blocktreeFirst 10 of 6,240 objects | 417,828,792 | 5.20% | 6,240
BlockTreeTermsReaderFirst 10 of 2,933 objects | 416,353,448 | 5.19% | 2,933
Total: 2 entries | 1,842,684,888 |   | 85,251
utilFirst 10 of 5,607 objects | 339,374,808 | 4.23% | 5,607
FixedBitsFirst 10 of 1,591 objects | 339,206,320 | 4.23% | 1,591
Total: 3 entries | 4,060,540,336 |   | 123,523
elasticsearchFirst 10 of 341,504 objects | 1,902,894,456 | 23.70% | 341,504
indicesFirst 10 of 158,478 objects | 803,637,920 | 10.01% | 158,478
IndicesQueryCache$ElasticsearchLRUQueryCacheAll 2 objects | 754,786,360 | 9.40% | 2
commonFirst 10 of 167,769 objects | 599,292,256 | 7.47% | 167,769
recyclerFirst 10 of 48 objects | 241,483,832 | 3.01% | 48
DequeRecyclerFirst 10 of 20 objects | 241,482,928 | 3.01% | 20
utilFirst 10 of 412 objects | 162,170,040 | 2.02% | 412
PageCacheRecyclerAll 2 objects | 161,910,216 | 2.02% | 2
cacheFirst 10 of 158,152 objects | 115,113,352 | 1.43% | 158,152
Cache$EntryFirst 10 of 158,138 objects | 105,003,680 | 1.31% | 158,138
Total: 3 entries | 518,767,224 |   | 158,612
indexFirst 10 of 3,285 objects | 288,017,520 | 3.59% | 3,285
IndexServiceFirst 10 of 61 objects | 199,436,968 | 2.48% | 61
searchFirst 10 of 781 objects | 170,820,352 | 2.13% | 781
SearchServiceAll 2 objects | 170,370,840 | 2.12% | 2
Total: 4 entries | 1,861,768,048 |   | 330,313
Total: 2 entries | 5,969,770,952 |   | 472,496
byte[]First 10 of 5,011 objects | 992,431,728 | 12.36% | 5,011
javaFirst 10 of 126,404 objects | 675,098,264 | 8.41% | 126,404
langFirst 10 of 101,112 objects | 664,370,256 | 8.28% | 101,112
ThreadFirst 10 of 338 objects | 653,106,128 | 8.14% | 338
ioFirst 10 of 11,102 objects | 384,385,200 | 4.79% | 11,102
nettyFirst 10 of 11,102 objects | 384,385,200 | 4.79% | 11,102
bufferFirst 10 of 7,420 objects | 379,483,704 | 4.73% | 7,420
PoolChunkFirst 10 of 129 objects | 354,472,384 | 4.42% | 129
Total: 4 entries | 8,021,804,712 |   | 615,168

Class histogram:
Class Name	Objects	Shallow Heap	Retained Heap
byte[]
All objects	3,481,815	3,936,315,992	>= 3,936,315,992
long[]
All objects	737,541	3,205,004,560	>= 3,205,004,560
org.apache.lucene.util.FixedBits
All objects	8,033	192,792	>= 2,444,792,464
java.util.HashMap
All objects	269,808	12,950,784	>= 1,721,627,000
java.util.HashMap$Node[]
All objects	101,124	11,569,696	>= 1,716,176,752
org.apache.lucene.index.SegmentReader
All objects	18,531	1,630,728	>= 1,676,086,048
org.apache.lucene.index.IndexWriter
All objects	151	31,408	>= 1,511,363,088
org.apache.lucene.index.ReaderPool
All objects	151	8,456	>= 1,472,692,568
org.apache.lucene.codecs.compressing.CompressingStoredFieldsReader
All objects	84,957	6,796,560	>= 1,386,803,376
org.apache.lucene.codecs.compressing.CompressingStoredFieldsReader$BlockState
All objects	84,957	4,757,592	>= 1,359,573,648
org.apache.lucene.util.BytesRef
All objects	636,774	15,282,576	>= 1,347,417,024
java.lang.Object[]
All objects	128,979	27,684,648	>= 1,225,122,704
java.util.WeakHashMap
All objects	52,307	2,510,736	>= 917,978,536
java.util.Collections$SynchronizedSet
All objects	47,912	1,149,888	>= 915,486,680
java.util.Collections$SetFromMap
All objects	44,749	1,073,976	>= 914,223,392
java.util.WeakHashMap$Entry[]
All objects	52,310	4,717,744	>= 913,484,528
java.util.HashMap$Node
All objects	882,020	28,224,640	>= 812,033,048
org.elasticsearch.indices.IndicesQueryCache$ElasticsearchLRUQueryCache
All 1 objects	1	88	>= 754,786,360
java.util.IdentityHashMap
All objects	1,032	41,280	>= 752,369,864
org.apache.lucene.search.LRUQueryCache$LeafCache
All objects	395	12,640	>= 752,069,536
org.apache.lucene.util.BitDocIdSet
All objects	2,347	56,328	>= 656,515,592
org.apache.lucene.util.FixedBitSet
All objects	2,478	59,472	>= 656,478,864
java.lang.Thread
All objects	339	40,680	>= 653,118,440
java.lang.ThreadLocal$ThreadLocalMap
All objects	673	16,152	>= 647,854,104
java.lang.ThreadLocal$ThreadLocalMap$Entry[]
All objects	673	4,461,968	>= 647,837,944

Total: 25 of 14,541 entries; 14,516 more
21,550,603	8,027,986,976	

Could this be a memory leak somewhere (or

**Provide logs (if relevant)**:
```
[2019-04-08T12:41:52,956][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][young][1688189][55387] duration [1.7s], collections [1]/[2.6s], total [1.7s]/[1.6h], memory [17.5gb]->[16.5gb]/[20gb],
 all_pools {[young] [6.7gb]->[48mb]/[0b]}{[survivor] [216mb]->[104mb]/[0b]}{[old] [10.6gb]->[16.3gb]/[20gb]}
[2019-04-08T12:41:53,003][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][1688189] overhead, spent [1.7s] collecting in the last [2.6s]
[2019-04-08T12:46:24,726][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][young][1688452][55400] duration [8.1s], collections [1]/[9s], total [8.1s]/[1.6h], memory [17.5gb]->[17.2gb]/[20gb], a
ll_pools {[young] [9.8gb]->[120mb]/[0b]}{[survivor] [216mb]->[16mb]/[0b]}{[old] [7.5gb]->[17.1gb]/[20gb]}
[2019-04-08T12:46:24,802][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][1688452] overhead, spent [8.1s] collecting in the last [9s]
[2019-04-08T12:50:55,367][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][young][1688715][55413] duration [7.2s], collections [1]/[7.7s], total [7.2s]/[1.6h], memory [17.5gb]->[16.9gb]/[20gb],
 all_pools {[young] [9.7gb]->[8mb]/[0b]}{[survivor] [192mb]->[24mb]/[0b]}{[old] [7.6gb]->[16.8gb]/[20gb]}
[2019-04-08T12:50:55,371][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][1688715] overhead, spent [7.2s] collecting in the last [7.7s]
[2019-04-08T12:53:28,724][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][young][1688860][55428] duration [7.8s], collections [1]/[8.8s], total [7.8s]/[1.6h], memory [17.3gb]->[16.8gb]/[20gb],
 all_pools {[young] [9.2gb]->[192mb]/[0b]}{[survivor] [160mb]->[0b]/[0b]}{[old] [7.9gb]->[16.7gb]/[20gb]}
[2019-04-08T12:53:28,743][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][1688860] overhead, spent [7.8s] collecting in the last [8.8s]
[2019-04-08T12:57:20,689][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][young][1689078][55442] duration [13s], collections [1]/[14s], total [13s]/[1.6h], memory [17.4gb]->[16.9gb]/[20gb], al
l_pools {[young] [9.6gb]->[64mb]/[0b]}{[survivor] [168mb]->[0b]/[0b]}{[old] [7.5gb]->[16.9gb]/[20gb]}
[2019-04-08T12:57:20,754][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][1689078] overhead, spent [13s] collecting in the last [14s]
[2019-04-08T13:04:17,007][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][young][1689485][55459] duration [8.8s], collections [1]/[9.2s], total [8.8s]/[1.6h], memory [17.6gb]->[17.2gb]/[20gb],
 all_pools {[young] [9.8gb]->[136mb]/[0b]}{[survivor] [208mb]->[0b]/[0b]}{[old] [7.6gb]->[17gb]/[20gb]}
[2019-04-08T13:04:17,104][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][1689485] overhead, spent [8.8s] collecting in the last [9.2s]
[2019-04-08T13:16:09,037][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][young][1690186][55482] duration [9.3s], collections [1]/[9.8s], total [9.3s]/[1.6h], memory [17.6gb]->[17.3gb]/[20gb],
 all_pools {[young] [9.9gb]->[112mb]/[0b]}{[survivor] [232mb]->[0b]/[0b]}{[old] [7.4gb]->[17.1gb]/[20gb]}
[2019-04-08T13:16:09,352][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][1690186] overhead, spent [9.3s] collecting in the last [9.8s]
[2019-04-08T13:20:53,549][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][young][1690468][55495] duration [1.8s], collections [1]/[1.9s], total [1.8s]/[1.6h], memory [17.6gb]->[16gb]/[20gb], a
ll_pools {[young] [9.9gb]->[80mb]/[0b]}{[survivor] [176mb]->[96mb]/[0b]}{[old] [7.5gb]->[15.9gb]/[20gb]}
[2019-04-08T13:20:53,578][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][1690468] overhead, spent [1.8s] collecting in the last [1.9s]
[2019-04-08T13:22:54,467][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][young][1690581][55505] duration [8.3s], collections [1]/[8.6s], total [8.3s]/[1.6h], memory [17.5gb]->[16.8gb]/[20gb],
 all_pools {[young] [9.4gb]->[144mb]/[0b]}{[survivor] [152mb]->[0b]/[0b]}{[old] [7.8gb]->[16.6gb]/[20gb]}
[2019-04-08T13:22:54,505][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][1690581] overhead, spent [8.3s] collecting in the last [8.6s]
[2019-04-08T13:25:49,479][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][young][1690755][55521] duration [1.2s], collections [1]/[1.4s], total [1.2s]/[1.6h], memory [17.6gb]->[13gb]/[20gb], a
ll_pools {[young] [10gb]->[88mb]/[0b]}{[survivor] [168mb]->[96mb]/[0b]}{[old] [7.4gb]->[12.9gb]/[20gb]}
[2019-04-08T13:25:49,499][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25.] [gc][1690755] overhead, spent [1.2s] collecting in the last [1.4s]
[2019-04-08T13:27:30,648][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25] [gc][young][1690842][55526] duration [14.1s], collections [1]/[14.9s], total [14.1s]/[1.6h], memory [17.5gb]->[17gb]/[20gb], all_pools {[young] [9.7gb]->[16mb]/[0b]}{[survivor] [112mb]->[0b]/[0b]}{[old] [7.6gb]->[17gb]/[20gb]}
[2019-04-08T13:27:30,649][WARN ][o.e.m.j.JvmGcMonitorService] [fmfe25] [gc][1690842] overhead, spent [14.1s] collecting in the last [14.9s]
java.lang.OutOfMemoryError: Java heap space
Dumping heap to /data/elasticsearch/heapdumps/java_pid82750.hprof ...
Heap dump file created [8530677372 bytes in 60.215 secs]
Terminating due to java.lang.OutOfMemoryError: Java heap space
```

**Comment by DaveCTurner** [comment]
Hi @bra-fsn, thanks for raising this. Yes, some heap dumps would be very useful. Can you upload them here please?

https://upload-staging.elstc.co/u/0079f0c8-8f70-43b6-bec4-f8f84e45541a

Certainly interested in the one you have analysed above, but if there are others that look different then please share them too.

**Comment by bra-fsn** [comment]
Thanks @DaveCTurner! I've uploaded the above dump.
Will attach some more later.

**Comment by DaveCTurner** [comment]
Thanks, got them. I've been looking at the first one you uploaded and I don't see anything terribly obvious yet. This is clearly a very busy node, and there's some hints that it needs more resources to deal with all the updates/deletes in your workload: for instance, the bulk of the 100s of MBs taken up by those `org.apache.lucene.index.IndexWriter`s looks to be tracking the live docs in each segment, which is only needed for segments containing deletes (which includes deletes due to updates).

The `org.elasticsearch.indices.IndicesQueryCache$ElasticsearchLRUQueryCache` also looks normal - by default it's limited to 10% of the heap, which appears to be 7.5GB, so 750MB seems right.

Below that there's a bunch of network buffers and some FSTs (10+MB). It all gets a bit deep into Lucene's internals about here, so I'm going to try and find someone more knowledgable to take a closer look.

**Comment by bra-fsn** [comment]
Yes, these nodes have work to do and indeed, there are a lot of updates here.
I'm not good in Lucene internals, so this is new to me. So if there is at least one deleted doc in a segment, all (latest?) docs are referenced in memory?
For us it essentially means all of the docs...
If this is true, having #40159 seems to be a must for us (so we can forcemerge big shards into smaller number of segments).
Thanks a lot for looking into this!

**Comment by DaveCTurner** [comment]
> all (latest, in case of updates) docs are referenced in memory?

Not quite, no, it's just a single bit for each document in the segment indicating whether it's live or not. One of the larger such segments on this node has nearly 40M documents, so that's about 5MB of heap, but I see 65 copies of this data which adds up to over 300MB. This corresponds with the enormous amount of heap retained by `org.apache.lucene.index.IndexWriter @ 0x8e322cd30` in your question. As far as I can tell we allocate another 5MB for this segment at each refresh, but we don't discard the old ones until all the searches that were using it have completed. There look to be quite a few scrolls open keeping these old copies around. Is there anything you can do to reduce the refreshes or to close these scrolls more quickly?

**Comment by bra-fsn** [comment]
Quite useful conversation, thank you very much for sharing details about the internals!
OMG, that sounds pretty bad. Do I get it right that when I open a scroll, all subsequent refreshes to all shards will lock these allocations until the scroll is cleared?
Or is it just for the allocations already made before the scroll was created?
As for the refreshes, I don't really see anything I can do. If I'm right they depend mainly on `refresh_interval` and the number of incoming modifications. We have already optimized this, so have higher `refresh_interval` where it's possible.
Are there anything else which can be done here?

As for keeping scroll contexts alive (which we really do, we scroll over selected docs in the large indices quite regularly), I'm also not sure.
We quite depend on this. Some can be rewritten with simple searches (maybe combined with `wait_for`s, so we know we won't hit already modified docs in the next search, which scroll automatically solves now).
Having read through the docs I know that it causes the segments to stick around, but the exact consequences of this was (and I guess still) not entirely clear.
If you could clarify this part in the doc that would be awesome. (currently it says we need more open file descriptors, but that's the minor problem here)

These are great points and help to understand what's happening here a lot.
But I guess if this would be the only problem here, I should have crashes every days and (luckily, not yet?) it's not the case.

I know I can see how many scroll contexts are open, but is it possible to query details about them? I mean for example how long they are open.
I will check this.

**Comment by atris** [comment]
What is your segment merging policy? Did you change the default from TieredMergePolicy?

As @DaveCTurner said, having open scroll contexts will cause old segments to lie around. Typically, TieredMergePolicy will ensure that you do not have more than 15% deleted documents, so IndexWriter instances tracking live docs will have a smaller footprint.

How frequent are your deletes? If you delete at a more or less constant rate over time with multiple scroll contexts open, you will eventually run out of memory due to Lucene's inability to merge old segments and get rid of the deleted documents. As Dave explained, deleted documents not purged yet also have the overhead corresponding to IndexWriters.

To verify this suspicion, you could synthetically trigger a larger delete rate (maybe in a test cluster) while keeping other parameters same. You would run into OOM faster if above is the case applicable to you.

To get details of scroll contexts currently applicable to an index, use node stats API:

`curl -XGET 'ES_HOST:ES_PORT/_nodes/stats/indices/search'`

**Comment by atris** [comment]
> To get details of scroll contexts currently applicable to an index, use node stats API:
> 
> `curl -XGET 'ES_HOST:ES_PORT/_nodes/stats/indices/search'`

Although, I dont recall offhand if this tells you how long a scroll context has been up for, so it might not really help.

**Comment by bra-fsn** [comment]
I know and use this, and sadly it doesn't tell it, just how many contexts are open currently.

**Comment by bra-fsn** [comment]
> What is your segment merging policy? Did you change the default from TieredMergePolicy?
Nope, it's the default.

> As @DaveCTurner said, having open scroll contexts will cause old segments to lie around. Typically, TieredMergePolicy will ensure that you do not have more than 15% deleted documents, so IndexWriter instances tracking live docs will have a smaller footprint.
I have 1.56B deleted docs out of 21.09B total currently (primaries).

> How frequent are your deletes? If you delete at a more or less constant rate over time with multiple scroll contexts open, you will eventually run out of memory due to Lucene's inability to merge old segments and get rid of the deleted documents. As Dave explained, deleted documents not purged yet also have the overhead corresponding to IndexWriters.
Quite frequent. I mean on a daily average I have (per second):
index: 4.92k
delete: 209
refresh: 918

Is Lucene smart enough to merge segments not involved in an open scroll context?
I mean these two questions:
1. (I guess this works because it's trivial) only keeps segments for involved indices (so if I do a search on index1, segments in other indices will be just merged)
2. if I open a scroll, all segments at that time will be preserved until the scroll context is alive. But will newly written segments be merged, or they will also be held until the context is destroyed?

> To verify this suspicion, you could synthetically trigger a larger delete rate (maybe in a test cluster) while keeping other parameters same. You would run into OOM faster if above is the case applicable to you.
I can try this. What parameter should contain the memory which will be growing faster than currently?

**Comment by atris** [comment]
> > How frequent are your deletes? If you delete at a more or less constant rate over time with multiple scroll contexts open, you will eventually run out of memory due to Lucene's inability to merge old segments and get rid of the deleted documents. As Dave explained, deleted documents not purged yet also have the overhead corresponding to IndexWriters.
> 
> Quite frequent. I mean on a daily average I have (per second):
> index: 4.92k
> delete: 209
> refresh: 918
> 

Ok, that might be the source of the problem.

> Is Lucene smart enough to merge segments not involved in an open scroll context?
> I mean these two questions:
> 
> 1. (I guess this works because it's trivial) only keeps segments for involved indices (so if I do a search on index1, segments in other indices will be just merged)

Yes.

> 2. if I open a scroll, all segments at that time will be preserved until the scroll context is alive. But will newly written segments be merged, or they will also be held until the context is destroyed?

I believe so. The semantic is that segments currently opened by an `IndexReader` cannot be merged, but unless you explicitly ask for a new `IndexReader` instance from your `Directory`, the changes being done by merging will not be visible to your executing query.

> > To verify this suspicion, you could synthetically trigger a larger delete rate (maybe in a test cluster) while keeping other parameters same. You would run into OOM faster if above is the case applicable to you.
> 
> I can try this. What parameter should contain the memory which will be growing faster than currently?

I am not sure I understand your question. If you increase the deletion rate of documents while keeping the number of open scroll contexts same (or maybe open a few more and keep them open to further stress the scenario), keep the ingestion rate constant, then you should hit OOM earlier. Even before OOM, check your consumed memory and it should increase at a faster rate than what happens in your current scenario.

**Comment by atris** [comment]
> I know and use this, and sadly it doesn't tell it, just how many contexts are open currently.

That is a limitation, and unless we have another way of getting this information, I would suggest addition of the same to Indices stats. @DaveCTurner Any thoughts on this?

**Comment by bra-fsn** [comment]
> > I can try this. What parameter should contain the memory which will be growing faster than currently?
> 
> I am not sure I understand your question. If you increase the deletion rate of documents while keeping the number of open scroll contexts same (or maybe open a few more and keep them open to further stress the scenario), keep the ingestion rate constant, then you should hit OOM earlier. Even before OOM, check your consumed memory and it should increase at a faster rate than what happens in your current scenario.

Is there a stats counter which shows the memory usage which I should monitor, which shows that it grows more steeply by having a bigger rate of deletes?
I monitor `http://localhost:9200/$INDEX/_stats` for every indices, but it doesn't show a sawtooth shape, which it should according to the above, no?
The `index_writer_memory_in_bytes` just moves with the number of documents in the index and doesn't fluctuate (this is one of the biggest indices which has a lot of deletes/updates):
![image](https://user-images.githubusercontent.com/820331/56429028-bbc57180-62c1-11e9-9125-9293d1708d8f.png)

If this is not the metric I should be watching, the question is what is it?
And if it's the right metric, the question is why it looks like this and not a sawtooth? :)

**Comment by bra-fsn** [comment]
> > I can try this. What parameter should contain the memory which will be growing faster than currently?
> 
> I am not sure I understand your question. If you increase the deletion rate of documents while keeping the number of open scroll contexts same (or maybe open a few more and keep them open to further stress the scenario), keep the ingestion rate constant, then you should hit OOM earlier. Even before OOM, check your consumed memory and it should increase at a faster rate than what happens in your current scenario.

Sorry, I mean should I watch for indices/segments/index_writer_memory_in_bytes in _nodes/stats and I should see the increased growth there?
And if this is the real cause of my issue, a sawtooth-like graph eventually? I'm asking this because while this fluctuates currently (seems normal), I can't see this shape.

**Comment by DaveCTurner** [comment]
> Having read through the docs I know that it causes the segments to stick around, but the exact consequences of this was (and I guess still) not entirely clear.

Yes, I opened #41375 to suggest a clarification, although this is not really my area of expertise so I wouldn't trust the accuracy of my wording until it's been reviewed.

> only keeps segments for involved indices (so if I do a search on index1, segments in other indices will be just merged)

Yes, there's no interaction between shards here.

> But will newly written segments be merged, or they will also be held until the context is destroyed?

No, it's just the existing segments that are held. Merging continues as normal. Segments held by a scroll are also merged as normal, except that they're not deleted after being merged while there's still an open context that needs them. From [the docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-scroll.html)

>Normally, the background merge process optimizes the index by merging together smaller segments to create new bigger segments, at which time the smaller segments are deleted. This process continues during scrolling, but an open search context prevents the old segments from being deleted while they are still in use. This is how Elasticsearch is able to return the results of the initial search request, regardless of subsequent changes to documents.

> I know I can see how many scroll contexts are open, but is it possible to query details about them?

No I can't see a way of doing this (unless you count taking a heap dump :grin:)

I do not think that increasing the _rate_ of deletions will exacerbate what we're seeing here. The big heap consumer I'm looking at here is held by the live docs arrays, which are a bit complicated to explain but I'll try to do so as follows. If a scroll is working on a segment containing deletes then it needs to hold 1-2 bits per document in that segment. But if there are multiple scrolls on that segment with the same sets of live docs then I think they share the data. If you want to increase the heap usage from live docs arrays, could you try the following:

1. find a segment with lots of docs (ideally, make a one-shard one-segment index using force-merge, it'll simplify the rest of this)
2. open a scroll involving this segment
3. delete a doc
4. refresh this index
5. GOTO 2

You should see heap usage increasing each time round, and then when you close each scroll you should see it drop again. I don't know that this is tracked in any stats, mind, so you might have to talk to the JVM to observe this, or to do it until it OOMs.

**Comment by bra-fsn** [comment]
> > I know and use this, and sadly it doesn't tell it, just how many contexts are open currently.
> 
> That is a limitation, and unless we have another way of getting this information, I would suggest addition of the same to Indices stats. @DaveCTurner Any thoughts on this?

Shouldn't this deserve a new stat endpoint (for querying the open scroll contexts)?
Just having the number may be not enough for debugging who/which holds these. It would be nice to have the client IP, the ID, the open time, the timeout and whatever else is available as well.

**Comment by DaveCTurner** [comment]
Since I think these OOMs are because your nodes don't have the resources needed to do what you're asking them to do, and not because of a bug, I think it best to close this issue and ask you to move this discussion over to the [discussion forum](https://discuss.elastic.co/c/elasticsearch). If it turns out that there is a bug here then we can of course re-open this.

**Comment by bra-fsn** [comment]
> > I know I can see how many scroll contexts are open, but is it possible to query details about them?
> 
> No I can't see a way of doing this (unless you count taking a heap dump grin)

It would be nice to have this feature (opened a feature request).
 
> I do not think that increasing the _rate_ of deletions will exacerbate what we're seeing here. The big heap consumer I'm looking at here is held by the live docs arrays, which are a bit complicated to explain but I'll try to do so as follows. If a scroll is working on a segment containing deletes then it needs to hold 1-2 bits per document in that segment. But if there are multiple scrolls on that segment with the same sets of live docs then I think they share the data. If you want to increase the heap usage from live docs arrays, could you try the following:

Do you see any possibilities to overcome this issue?

> 
>     1. find a segment with lots of docs (ideally, make a one-shard one-segment index using force-merge, it'll simplify the rest of this)
> 
>     2. open a scroll involving this segment
> 
>     3. delete a doc
> 
>     4. refresh this index
> 
>     5. GOTO 2
> 
> 
> You should see heap usage increasing each time round, and then when you close each scroll you should see it drop again. I don't know that this is tracked in any stats, mind, so you might have to talk to the JVM to observe this, or to do it until it OOMs.

I think it's vital to have this in the stats. I have graphs from them and currently nothing indicates that there is a problem here.
According to the above, shouldn't index_writer_memory_in_bytes contain this?

**Comment by atris** [comment]
> I do not think that increasing the _rate_ of deletions will exacerbate what we're seeing here. The big heap consumer I'm looking at here is held by the live docs arrays, which are a bit complicated to explain but I'll try to do so as follows. If a scroll is working on a segment containing deletes then it needs to hold 1-2 bits per document in that segment. But if there are multiple scrolls on that segment with the same sets of live docs then I think they share the data. If you want to increase the heap usage from live docs arrays, could you try the following:

Yes, multiple scrolls would share the same `IndexWriter` for live docs tracking for same segment.

I am curious to understand why increase the rate of deletions keeping everything else constant would not aggravate this scenario. If a large segment was opened by multiple scrolls and the deletion rate for that segment grew significantly, would the live docs array not grow proportionately since you cannot merge the segment yet?


> 1. find a segment with lots of docs (ideally, make a one-shard one-segment index using force-merge, it'll simplify the rest of this)
> 2. open a scroll involving this segment
> 3. delete a doc
> 4. refresh this index
> 5. GOTO 2
> 

Would it not be a better way to aggravate if each scroll was opened on a different segment, since that means that there would be different live docs array per scroll?

**Comment by bra-fsn** [comment]
BTW, open scrolls per node:
![image](https://user-images.githubusercontent.com/820331/56437080-6a77ab00-62de-11e9-84c7-892c0f60bea3.png)

and index_writer_memory:
![image](https://user-images.githubusercontent.com/820331/56437116-80856b80-62de-11e9-983b-0c151a3f39c8.png)

Az index_writer_memory cluster wide:
![image](https://user-images.githubusercontent.com/820331/56444244-37421580-62f8-11e9-8176-3fafda867700.png)

These look pretty fine I would say. So is that index_writer_memory doesn't account all of its memory usage a bug, or there should be a new metric introduced?
Could you please help to find where that memory is in the stats?

**Comment by atris** [comment]
@bra-fsn The portion of `index_writer_memory` graph that you showed is too less of data to really comment on, but from what's available, isnt that graph following the same pattern as `index_writer_memory_cluster`?

Can you highlight all the three graphs for a smaller (and the same) time window? That would help corraborate patterns across the three.

**Comment by bra-fsn** [comment]
@atris: here are the graphs (sadly some missing interval has slipped in):
This is index_memory_in_bytes cluster wide:
![image](https://user-images.githubusercontent.com/820331/56607762-94e0a580-6609-11e9-9b8f-5a694fb0e9c5.png)
This is index_writer_memory_in_bytes per node:
![image](https://user-images.githubusercontent.com/820331/56607874-d1ac9c80-6609-11e9-8860-4043640eb5ff.png)
This is number of scrolls per node:
![image](https://user-images.githubusercontent.com/820331/56607823-b8a3eb80-6609-11e9-988e-61e3c8f4ce16.png)
Scrolls cluster wide:
![image](https://user-images.githubusercontent.com/820331/56607928-f274f200-6609-11e9-9693-ebfa4c3cce1c.png)

According to this, I would say either the assumption is wrong, or not all memory is accounted in index_writer_memory.

**Comment by bra-fsn** [comment]
@DaveCTurner, @atris: so index_writer_memory per node (according to elastic) tops at around 20-50 MiBs and doesn't show a pattern of accumulation, it just surges with the traffic rate.
So how could I prove that the assumption about IndexWriter eating all of the memory is true? How can I measure this?

**Comment by DaveCTurner** [comment]
Hi @bra-fsn, repeating https://github.com/elastic/elasticsearch/issues/41337#issuecomment-484926863 could you continue this conversation on the [discussion forum](https://discuss.elastic.co/c/elasticsearch)? I'll try and answer you over there, but GitHub isn't the right place for this discussion.
